src/codegen/compile.ml

```diff
@@ -3697,6 +3727,148 @@ module IC = struct
 
 end (* IC *)
 
+module StableMem = struct
+
+  let version = Int32.of_int 0
+
+  let register_globals env =
+    (* size (in pages) *)
+    E.add_global32 env "__stablemem_size" Mutable 0l
+
+  let get_mem_size env =
+    G.i (GlobalGet (nr (E.get_global env "__stablemem_size")))
+  let set_mem_size env =
+    G.i (GlobalSet (nr (E.get_global env "__stablemem_size")))
+
+  (* stable memory bounds check *)
+  let _guard env =
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code1 env "__stablemem_guard"
+        ("offset", I32Type) []
+        (fun env get_offset ->
+          get_offset ^^
+          compile_unboxed_const (Int32.of_int page_size_bits) ^^
+          G.i (Binary (Wasm.Values.I32 I32Op.ShrU)) ^^
+          get_mem_size env ^^
+          G.i (Compare (Wasm.Values.I32 I32Op.LtU)) ^^
+          E.else_trap_with env "StableMemory offset out of bounds")
+    | _ -> assert false
+
+  let _guard_range env =
+    (* is there something more efficient? *)
+    match E.mode env with
+    | Flags.ICMode | Flags.RefMode ->
+      Func.share_code2 env "__stablemem_guard"
+        (("offset", I32Type), ("size", I32Type)) []
+        (fun env get_offset get_size ->
+          get_offset ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
+          get_size ^^ G.i (Convert (Wasm.Values.I64 I64Op.ExtendUI32)) ^^
+          G.i (Binary (Wasm.Values.I64 I64Op.Add)) ^^
+          get_mem_size env ^^
+          compile_const_64 (Int64.of_int page_size_bits) ^^
+          G.i (Binary (Wasm.Values.I64 I64Op.Shl)) ^^
```

Comment:
Is there a reason why you're not using the right shift here?

Reply:
It seems more complicated actually (see version using Shr in comments, but perhaps you had something better in mind)?